### PR TITLE
Implement andreas's ghc-lib-gen speedup patch

### PR DESCRIPTION
**Pull request by shayne-fletcher**
This PR implements @aherrmann-da 's very cool speedup hack for `ghc -M` as provided in https://github.com/digital-asset/daml/pull/14076 for all builds.

This seems to provide a ~1/3 reduction in build time for a `stack runhaskell ... CI.hs`.

Pushing the lexer and parser code gen steps to the user doesn't seem to increase client build times too much.

Lastly, this approach enables removing quite a lot of now unnecessary code which is quite a compelling reason to adopt it in itself.

```
andreas:
      380.06 real      1093.73 user       365.33 sys
user + sys: 1459.06

master:
      574.09 real      1217.91 user       427.20 sys
user + sys: 1645.11

relative reductions:
  - wall time: 33%
  - cpu time : 11%
```

cc @akrmn @aherrmann-da @cocreature @garyverhaegen-da @sofiafaro-da

### CODE REVIEW

**Comment by shayne-fletcher** [comment]
Note as discussed in https://github.com/digital-asset/ghc-lib/pull/377, the DA builds can be expected to fail at this time. The remaining builds have been verified however.

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 1]
```diff
@@ -365,7 +374,8 @@ calcParserModules ghcFlavor = do
       strippedModulePaths = foldl
         (\acc p -> map (replace (p ++ "/") "") acc)
         modulePaths
-        hsSrcDirs
+        --hsSrcDirs
```
Comment (akrmn) [inline ghc-lib-gen/src/Ghclibgen.hs:377]:
```suggestion
```

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs]: thanks!

**Review by akrmn — APPROVED** [review]
Thank you! Sorry I took a while, I wanted to make sure I understood the changes. I'll take care of rerunning CI once https://github.com/digital-asset/ghc-lib/pull/380 is merged.

**Comment by shayne-fletcher** [comment]
> Thank you! Sorry I took a while, I wanted to make sure I understood the changes. I'll take care of rerunning CI once #380 is merged.

i appreciate you going to the trouble. thank-you!

**Comment by akrmn** [comment]
thanks @garyverhaegen-da!

**Comment by aherrmann-da** [comment]
@shayne-fletcher Thanks a lot for generalizing this onto ghc-lib `master`! I've adapted the Daml PR to use this updated version of ghc-lib and it produces the same benefits.

**Pull request merged**